src/node_sockaddr-inl.h

```diff
@@ -0,0 +1,170 @@
+#ifndef SRC_NODE_SOCKADDR_INL_H_
+#define SRC_NODE_SOCKADDR_INL_H_
+
+#if defined(NODE_WANT_INTERNALS) && NODE_WANT_INTERNALS
+
+#include "node.h"
+#include "node_internals.h"
+#include "node_sockaddr.h"
+#include "util-inl.h"
+
+#include <string>
+
+namespace node {
+
+static constexpr uint32_t kLabelMask = 0xFFFFF;
+
+inline void hash_combine(size_t* seed) { }
+
+template <typename T, typename... Args>
+inline void hash_combine(size_t* seed, const T& value, Args... rest) {
+    *seed ^= std::hash<T>{}(value) + 0x9e3779b9 + (*seed << 6) + (*seed >> 2);
+    hash_combine(seed, rest...);
+}
+
+bool SocketAddress::is_numeric_host(const char* hostname) {
+  return is_numeric_host(hostname, AF_INET) ||
+         is_numeric_host(hostname, AF_INET6);
+}
+
+bool SocketAddress::is_numeric_host(const char* hostname, int family) {
+  in6_addr dst;
+  return inet_pton(family, hostname, &dst) == 1;
+}
+
+int SocketAddress::GetPort(const sockaddr* addr) {
+  CHECK(addr->sa_family == AF_INET || addr->sa_family == AF_INET6);
+  return ntohs(addr->sa_family == AF_INET ?
+      reinterpret_cast<const sockaddr_in*>(addr)->sin_port :
+      reinterpret_cast<const sockaddr_in6*>(addr)->sin6_port);
+}
+
+int SocketAddress::GetPort(const sockaddr_storage* addr) {
+  return GetPort(reinterpret_cast<const sockaddr*>(addr));
+}
+
+std::string SocketAddress::GetAddress(const sockaddr* addr) {
+  CHECK(addr->sa_family == AF_INET || addr->sa_family == AF_INET6);
+  char host[INET6_ADDRSTRLEN];
+  const void* src = addr->sa_family == AF_INET ?
+      static_cast<const void*>(
```

Comment:
Nit: this static cast could go around the ternary to make this a bit shorter.

Reply:
Unfortunately not, because the types for sin_addr and sin6_addr are different. Moving to 

```c
  const void* src = static_cast<const void*>(
    addr->sa_family == AF_INET ?
        &(reinterpret_cast<const sockaddr_in*>(addr)->sin_addr) :
        &(reinterpret_cast<const sockaddr_in6*>(addr)->sin6_addr));
```

results in a compile error:

```
C:\Users\jasne\Projects\node\src\node_sockaddr-inl.h(52,66): error C2446: ':': no conversion from 'const IN6_ADDR *' to 'const IN_ADDR *' [C:\Users\jasne\Projects\node\libnode.vcxproj]
```

We might be able to simplify it in other ways but the static casts here are correct.